drivers/kscan/kscan_xpt2046.c

```diff
@@ -0,0 +1,276 @@
+	[0] = START | CHANNEL(CH_Z1) | POWER_ON,
+	[2] = START | CHANNEL(CH_Z2) | POWER_ON,
+	[4] = START | CHANNEL(CH_X) | POWER_ON,
+	[6] = START | CHANNEL(CH_Y) | POWER_OFF,
+};
+
+static void xpt2046_isr_handler(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct xpt2046_data *data = CONTAINER_OF(cb, struct xpt2046_data, int_gpio_cb);
+	const struct xpt2046_config *config = data->dev->config;
+
+	gpio_remove_callback(config->int_gpio.port, &data->int_gpio_cb);
+	k_work_submit(&data->work);
+}
+
+static int xpt2046_read_and_cumulate(const struct spi_dt_spec *bus, const struct spi_buf_set *tx,
+				     const struct spi_buf_set *rx, struct measurement *meas)
+{
+	int ret = spi_transceive_dt(bus, tx, rx);
+
+	if (ret) {
+		LOG_ERR("spi_transceive() %d\n", ret);
+		return ret;
+	}
+
+	uint8_t *buf = rx->buffers->buf;
+
+	meas->z += CONVERT_U16(buf, 1) + 4096 - CONVERT_U16(buf, 3);
+	meas->x += CONVERT_U16(buf, 5);
+	meas->y += CONVERT_U16(buf, 7);
+	return 0;
+}
+
+static void xpt2046_release_handler(struct k_work *kw)
+{
+	struct k_work_delayable *dw = k_work_delayable_from_work(kw);
+	struct xpt2046_data *data = CONTAINER_OF(dw, struct xpt2046_data, dwork);
+
+	if (data->pressed && data->enabled) {
+		data->pressed = false;
+		data->callback(data->dev, data->last_y, data->last_y, false);
+	}
+}
+
+static void xpt2046_work_handler(struct k_work *kw)
+{
+	struct xpt2046_data *data = CONTAINER_OF(kw, struct xpt2046_data, work);
+	struct xpt2046_config *config = (struct xpt2046_config *)data->dev->config;
+
+	const struct spi_buf txb = {.buf = tbuf, .len = sizeof(tbuf)};
+	const struct spi_buf rxb = {.buf = data->rbuf, .len = sizeof(data->rbuf)};
+	const struct spi_buf_set tx_bufs = {.buffers = &txb, .count = 1};
+	const struct spi_buf_set rx_bufs = {.buffers = &rxb, .count = 1};
+
+	/* Measure 3 times and calculate average */
+	int_fast16_t rounds = config->reads;
+	struct measurement meas = {0};
+
+	for (int i = 0; i < rounds; i++) {
+		if (xpt2046_read_and_cumulate(&config->bus, &tx_bufs, &rx_bufs, &meas) != 0) {
+			return;
+		}
+	}
+	meas.x /= rounds;
+	meas.y /= rounds;
+	meas.z /= rounds;
+
+	/* Calculate Xp = M * Xt + C using fixed point aritchmetics, where
+	 * Xp is the point in screen coordinates, Xt is the touch coordinates.
+	 * Use signed int32_t for calculation to ensure that we cover the roll-over to negative
+	 * values and return zero instead.
+	 */
+	int32_t mx = (config->screen_size_x << 16) / (config->max_x - config->min_x);
+	int32_t cx = (config->screen_size_x << 16) - mx * config->max_x;
+	int32_t x = mx * meas.x + cx;
+
+	x = (x < 0 ? 0 : x) >> 16;
+
+	int32_t my = (config->screen_size_y << 16) / (config->max_y - config->min_y);
+	int32_t cy = (config->screen_size_y << 16) - my * config->max_y;
+	int32_t y = my * meas.y + cy;
+
+	y = (y < 0 ? 0 : y) >> 16;
+
+	bool pressed = meas.z > config->threshold;
+
+	/* Don't send any other than "pressed" events.
+	 * releasing seem to cause just random noise
```

Comment:
I refactored the driver a bit so that as long as the PENIRQ is down, assume that user is still pressing. And keep checking this on a worker-queue event. Once the pressure is released, send the "released" event.

This seem to work fine for LVGL.. with one exception: Dragging does not work, because I do not re-read values from SPI. I cannot decide, whether to spend enormous amount of SPI traffic to support dragging, or just be stable with button presses.